zookeeper-specifications/zab-1.0/ZabWithFLE.tla

```diff
@@ -0,0 +1,1020 @@
+                                   mepoch |-> acceptedEpoch'[i]])
+        /\ UNCHANGED <<state, currentEpoch, lastZxid, zabState, history, commitIndex, electionVarsZ, leadingVoteSet, learners, ackeRecv, ackldRecv, 
+                       forwarding, ackIndex, currentCounter, sendCounter, committedIndex, committedCounter, 
+                       tempVarsZ, followerVarsZ, verifyVarsZ, electionMsgs, idTable>>
+
+(* In phase f12, follower receives NEWEPOCH. If e' > f.p, then follower sends ACK-E back,
+   and ACK-E contains f.a and lastZxid to let leader judge whether it is the latest.
+   After handling NEWEPOCH, follower's zabState turns to SYNCHRONIZATION. *)
+FollowerHandleLEADERINFO(i, j) ==
+        /\ state[i] = FOLLOWING
+        /\ msgs[j][i] /= << >>
+        /\ msgs[j][i][1].mtype = LEADERINFO
+        /\ LET msg     == msgs[j][i][1]
+               infoOk  == j = leaderAddr[i]
+               epochOk == /\ infoOk
+                          /\ msg.mepoch >= acceptedEpoch[i]
+               correct == /\ epochOk
+                          /\ zabState[i] = DISCOVERY
+           IN /\ infoOk
+              /\ \/ /\ epochOk                      \* 1. Normal case
+                    /\ \/ /\ correct
+                          /\ acceptedEpoch' = [acceptedEpoch EXCEPT ![i] = msg.mepoch]
+                          /\ Reply(i, j, [mtype      |-> ACKEPOCH,
+                                          mepoch     |-> msg.mepoch,
+                                          mlastEpoch |-> currentEpoch[i],
+                                          mlastZxid  |-> lastZxid[i]])
+                          /\ cepochSent' = [cepochSent EXCEPT ![i] = TRUE]
+                          /\ UNCHANGED inherentViolated
+                       \/ /\ ~correct
+                          /\ PrintT("Exception: Condition correct is false in FollowerHandleLEADERINFO(" \o ToString(i) \o ", " \o ToString(j) \o ").")
+                          /\ inherentViolated' = TRUE
+                          /\ Discard(j, i)
+                          /\ UNCHANGED <<acceptedEpoch, cepochSent>>
+                    /\ zabState' = [zabState EXCEPT ![i] = IF zabState[i] = DISCOVERY THEN SYNCHRONIZATION
+                                                                                      ELSE zabState[i]]
+                    /\ UNCHANGED <<varsL, leaderAddr>>
+                 \/ /\ ~epochOk                    \* 2. Abnormal case - go back to election
+                    /\ FollowerShutdown(i)
+                    /\ Clean(i, j)
+                    /\ UNCHANGED <<acceptedEpoch, cepochSent, inherentViolated>>
+        /\ UNCHANGED <<history, commitIndex, learners, cepochRecv, ackeRecv, ackldRecv, forwarding, ackIndex, currentCounter, sendCounter, committedIndex,
+                       committedCounter, tempVarsZ, synced, proposalMsgsLog, epochLeader>>
+
+\* Abstraction of actions making follower synced with leader before leader sending NEWLEADER.
+subRECOVERYSYNC(i, j) ==
```

Comment:
Consider sharing some of this code with the RECOVERYSYNC action to avoid repetition.